### ISSUE TITLE
getSemanticDiagnostics() returns type errors even when the CLI `tsc` passes

### ISSUE THREAD

**Issue by NickHeiner** [dsherret/ts-morph#925]
I have a sample repo in which calling `getSemanticDiagnostics()` returns type errors, but invoking `tsc` on the command line for the same TS input code passes.

@ts-morph/bootstrap@0.8.0
@ts-morph/common@0.7.3
Node: v12.14.0

**To Reproduce**

1. Clone https://github.com/NickHeiner/ts-inconsistent-errors-demo.
2. Run `yarn typecheck`. Observe that it passes.
3. Run `node ts-morph-demo.js`. Observe that a type error is logged.

**Expected behavior**

Programmatic `ts-morph` usage matches results from the CLI.

**Notes**

As I was trying to maximally simplify the example, I noticed:

1. If all code is in a single monorepo package, instead of being split across monorepo packages, the inconsistency disappears.
2. If I make this change, the inconsistency disappears.
```diff
- export default function withFocusPath<P = any>(
+ export default function withFocusPath<P>(
```

I tried to also do a repro where I called `typescript` directly, to exonerate ts-morph, but it was too complicated to get that working in the time allotted. This underscores the value of ts-morph for providing a simpler API. :smile:

**Comment by NickHeiner** [comment]
In a separate context, I did something similar to this with TS 3.9.7, and I didn't observe the inconsistency.

**Comment by NickHeiner** [comment]
Hey @dsherret – could we possibly chat offline about how I can help with this? Can you DM me at twitter.com/nickheiner?

**Comment by dsherret** [comment]
@NickHeiner sorry for my delay. This past while has been one of the busiest in my life.

Thanks for the great reproduction steps!

**File Reading Differences**

I looked into this a bit and these are the files that ts-morph reads:

```
> node .\ts-morph-demo.js
V:/ts-inconsistent-errors-demo/packages/dep/index.d.ts
V:/ts-inconsistent-errors-demo/node_modules/react/package.json
V:/ts-inconsistent-errors-demo/node_modules/@types/react/package.json
V:/ts-inconsistent-errors-demo/node_modules/dep/package.json
V:/ts-inconsistent-errors-demo/node_modules/@types/react/index.d.ts
V:/ts-inconsistent-errors-demo/node_modules/@types/react/global.d.ts
V:/ts-inconsistent-errors-demo/node_modules/csstype/package.json
V:/ts-inconsistent-errors-demo/node_modules/prop-types/package.json
V:/ts-inconsistent-errors-demo/node_modules/@types/prop-types/package.json
V:/ts-inconsistent-errors-demo/node_modules/csstype/index.d.ts
V:/ts-inconsistent-errors-demo/node_modules/@types/prop-types/index.d.ts
V:/ts-inconsistent-errors-demo/packages/main/index.tsx
V:/ts-inconsistent-errors-demo/node_modules/dep/index.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2016.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es5.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.core.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.collection.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.iterable.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.symbol.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.generator.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.promise.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.proxy.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.reflect.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.symbol.wellknown.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2016.array.include.d.ts
```

And these are the ones tsc reads:

```
> yarn typecheck
V:/ts-inconsistent-errors-demo/packages/dep/index.d.ts
V:/ts-inconsistent-errors-demo/node_modules/react/package.json
V:/ts-inconsistent-errors-demo/node_modules/@types/react/package.json
V:/ts-inconsistent-errors-demo/node_modules/dep/package.json
V:/ts-inconsistent-errors-demo/node_modules/@types/react/index.d.ts
V:/ts-inconsistent-errors-demo/node_modules/@types/react/global.d.ts
V:/ts-inconsistent-errors-demo/node_modules/csstype/package.json
V:/ts-inconsistent-errors-demo/node_modules/prop-types/package.json
V:/ts-inconsistent-errors-demo/node_modules/@types/prop-types/package.json
V:/ts-inconsistent-errors-demo/node_modules/csstype/index.d.ts
V:/ts-inconsistent-errors-demo/node_modules/@types/prop-types/index.d.ts
V:/ts-inconsistent-errors-demo/packages/main/index.tsx
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2016.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es5.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.core.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.collection.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.iterable.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.symbol.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.generator.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.promise.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.proxy.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.reflect.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2015.symbol.wellknown.d.ts
V:/ts-inconsistent-errors-demo/node_modules/typescript/lib/lib.es2016.array.include.d.ts
```

So the difference is that ts-morph also reads this file, but I'm not sure why:

```
V:/ts-inconsistent-errors-demo/node_modules/dep/index.d.ts
```

**Root Names & Compiler Options Differences**

Here is ts-morph:

```
  rootNames: [
    'V:/ts-inconsistent-errors-demo/packages/dep/index.d.ts',
    'V:/ts-inconsistent-errors-demo/packages/main/index.tsx'
  ],
  options: {
    allowJs: true,
    esModuleInterop: true,
    isolatedModules: true,
    noUnusedLocals: true,
    jsx: 2,
    lib: [ 'lib.es2016.d.ts' ],
    module: 1,
    moduleResolution: 2,
    noEmit: true,
    strict: true,
    target: 2,
    resolveJsonModule: true,
    types: [],
    configFilePath: 'V:/ts-inconsistent-errors-demo/tsconfig.json'
  }
```

And here is tsc:

```
  rootNames: [
    'V:/ts-inconsistent-errors-demo/packages/dep/index.d.ts',
    'V:/ts-inconsistent-errors-demo/packages/main/index.tsx'
  ],
  options: {
    allowJs: true,
    esModuleInterop: true,
    isolatedModules: true,
    noUnusedLocals: true,
    jsx: 2,
    lib: [ 'lib.es2016.d.ts' ],
    module: 1,
    moduleResolution: 2,
    noEmit: true,
    strict: true,
    target: 2,
    resolveJsonModule: true,
    types: [],
    project: 'V:/ts-inconsistent-errors-demo',
    configFilePath: 'V:/ts-inconsistent-errors-demo/tsconfig.json'
  }
```

So the only difference was the `project` option. I tried added `project` manually by specifying....

```ts
  const project = await createProject({
    tsConfigFilePath: require.resolve('./tsconfig.json'),
    compilerOptions: {
      project: "V:/ts-inconsistent-errors-demo"
    }
  });
```

...but unfortunately that didn't do anything.

**Providing custom compiler host**

I tried providing a compiler host created by `ts.createCompilerHost` to `createProgram`:

```ts
const host = ts.createCompilerHost(project.compilerOptions.get(), true);
const diagnostics = ts.getPreEmitDiagnostics(project.createProgram({ host }));
```

...and the issue no longer happens when providing `skipFileDependencyResolution: true` to `createProject`, but it does happen when it's `false`. So I think there are two things to figure out here:

1. When setting `skipFileDependencyResolution` to `true`, why does the ts-morph compiler host cause this error and providing a compiler host created via `ts.createCompilerHost` does not?
2. Why does it not work at all when `skipFileDependencyResolution` is `false`?

**Solving Question 1**

I've created a host via `const host = ts.createCompilerHost(compilerOptions.get());` and am replacing all the methods on the ts-morph compiler host with it. Issue still persists after replacing all the methods, so it must be a missing method. Compared the keys between the two objects. After trying out some things I noticed that the issue does not occur after providing a `realpath` method on the compiler host.

**Solving Question 2**

Issue no longer occurs after solving question 1. Ok, probably occurs because when calling `createProgram` it's reusing the old program and the old program didn't have a host with `realpath` defined. Makes sense. It's probably reusing some stuff that was improperly resolved.

**Debugging**

In terms of debugging this, the way I do it is via the node_modules/typescript/lib/typescript.js file.

**Comment by dsherret** [comment]
Published in ts-morph/common 0.7.4. Thanks again for the steps and sorry for not getting back to you sooner.

**Comment by NickHeiner** [comment]
Great! I'm excited to try this out when I get back to work. Thanks for the detailed writeup.

And absolutely no worries about the timing; you don't owe anyone anything with this project, and you certainly don't owe me. :smile:
